tests/generative/variable_strategies.py

```diff
@@ -183,12 +200,84 @@ def subtract(draw, type_, frame):
     def multiply(draw, type_, frame):
         return draw(binary_operation(type_, frame, Function.Multiply))
 
+    @st.composite
+    def date_add_years(draw, type_, frame):
+        return draw(
+            binary_operation_with_types(type_, int, frame, Function.DateAddYears)
+        )
+
+    @st.composite
+    def date_add_months(draw, type_, frame):
+        return draw(
+            binary_operation_with_types(type_, int, frame, Function.DateAddMonths)
+        )
+
+    @st.composite
+    def date_add_days(draw, type_, frame):
+        return draw(
+            binary_operation_with_types(type_, int, frame, Function.DateAddDays)
+        )
+
+    @st.composite
+    def date_difference_in_years(draw, type_, frame):
+        return draw(
+            binary_operation(datetime.date, frame, Function.DateDifferenceInYears)
+        )
+
+    @st.composite
+    def date_difference_in_months(draw, type_, frame):
+        return draw(
+            binary_operation(datetime.date, frame, Function.DateDifferenceInMonths)
+        )
+
+    @st.composite
+    def date_difference_in_days(draw, type_, frame):
+        return draw(
+            binary_operation(datetime.date, frame, Function.DateDifferenceInDays)
+        )
+
     @st.composite
     def binary_operation(draw, type_, frame, operator_func):
         # A strategy for operations that take lhs and rhs arguments of the
         # same type
-        lhs = draw(series(type_, draw(one_row_per_patient_frame_or(frame))))
-        rhs = draw(series(type_, draw(one_row_per_patient_frame_or(frame))))
+        return draw(binary_operation_with_types(type_, type_, frame, operator_func))
+
+    @st.composite
+    def binary_operation_with_types(draw, lhs_type, rhs_type, frame, operator_func):
+        # A strategy for operations that take lhs and rhs arguments with specified lhs
+        # and rhs types (which may be different)
+
+        # A binary operation has 2 inputs, which are
+        # 1) A series drawn from the specified frame
+        # 2) one of:
+        #    a) A series drawn from the specified frame
+        #    b) A series drawn from a one-row-per-patient-frame
+        #    c) A Value
+
+        # Define other_frame and other_series (#2 above)
+        # pick either a one-row-per-patient-frame or this frame
```

Comment:
This comment refers to `other_series`, which I think is correct and clearer than `other_strategy` which is used in the code.

Reply:
But other_series could be a value, which is confusing.

I've just rewritten this function to pick the input type (series or value) first, which I think makes it clearer, and still does the right thing:

```
    @st.composite
    def binary_operation_with_types(draw, lhs_type, rhs_type, frame, operator_func):
        # A strategy for operations that take lhs and rhs arguments with specified lhs
        # and rhs types (which may be different)

        # A binary operation has 2 inputs, which are
        # 1) A series drawn from the specified frame
        # 2) one of:
        #    a) A series drawn from the specified frame
        #    b) A series drawn from a one-row-per-patient-frame
        #    c) A Value

        # first pick a input strategy, either a value or a series
        other_input = draw(st.sampled_from([value, series]))
        if other_input == value:
            # A value strategy ignores its frame input, so we can give it anything here
            other_frame = None
        else:
            # The input is a series, so now pick a frame for it to be drawn from
            # The other frame will either be a new one-row-per-patient-frame or this frame
            other_frame = draw(st.one_of(one_row_per_patient_frame(), st.just(frame)))

        # Pick the order of the lhs and rhs inputs built from the two frames and
        # associated strategies
        lhs_frame, lhs_input, rhs_frame, rhs_input = draw(
            st.sampled_from(
                [
                    (frame, series, other_frame, other_input),
                    (other_frame, other_input, frame, series)
                ]
            )
        )
        lhs = draw(lhs_input(lhs_type, lhs_frame))
        rhs = draw(rhs_input(rhs_type, rhs_frame))

        return operator_func(lhs, rhs)
```

Reply:
I think that's clear and correct and I'd be happy for you to merge it as-is.

But of course I have some thoughts...

* A `Value` _is_ a `Series`, it's just not one that can be created by the `series` strategy. This is a point where our generation code does not align directly with the type structure of the query model because the query model can express stuff which is both unrealistic and unhelpful. We have similar non-generality in the `Filter`-construction strategy.
* While it's correct that `Value` will ignore its input frame, there are plenty of other places where we knowingly pass ignored values to series strategies (for example type to `eq()` and frame to `count()`). I'm not sure that avoiding doing that here is worth the extra branching that it requires (but I'm quite happy to be wrong).

Reply:
So it can be reduced to this?
```
        # first pick an "other" input series (i.e. #2 above), either a value series 
        # or a series drawn from a frame
        other_series = draw(st.sampled_from([value, series]))
        # Now pick a frame for the series to be drawn from
        # The other frame will either be a new one-row-per-patient-frame or this frame
        # (Note if the other_series is a value, the frame will be ignored)
        other_frame = draw(st.one_of(one_row_per_patient_frame(), st.just(frame)))
```
Also, yes, of course, I was thinking of value and series as two different things.